src/util/sandbox/test_sandbox.c

```diff
@@ -0,0 +1,366 @@
+    } else if ( FD_UNLIKELY ( WIFSTOPPED( wstatus ) ) ) {
+      FD_LOG_WARNING(( "expected fd_sandbox_setup_netns to get killed by signal but " "stopped by signal %s", strsignal(WSTOPSIG(wstatus)) ));
+      return -1;
+    } else {
+      FD_LOG_WARNING(( "unexpected termination: wstatus(%d)", wstatus ));
+      return -1;
+    }
+    return -1;
+
+  } else { // child
+    fd_sandbox_seccomp( NULL );
+
+    // The call to execl is expected to have the kernel send a
+    // SIGSYS to the process group. The branch should not execute.
+    if ( execl( "/bin/true", "" ) ) {
+      // This could change if we ever switch stance on seccomp violations.
+      FD_LOG_WARNING(( "execl: should not have been executed %s", strerror(errno) ));
+      exit(0);
+    }
+
+    exit(0);
+  }
+
+return -1;
+}
+
+/* test_setup_netns_null ensures that, after calling `fd_sandbox_setup_mountns`,
+   the root mount should be empty and belong to root.
+*/
+int
+test_setup_netns_null() {
+  pid_t pid, w;
+
+  if (( pid = fork() )) { // parent
+    int wstatus;
+    w = waitpid(pid, &wstatus, WUNTRACED);
+    if ( FD_UNLIKELY ( w == -1 ) ) {
+      FD_LOG_WARNING(( "waitpid: %s", strerror(errno) ));
+      return -1;
+    }
+    if ( FD_LIKELY ( !wstatus ) ) {
+      return 0;
+    } else if ( FD_UNLIKELY ( WIFEXITED( wstatus ) ) ) {
+      FD_LOG_WARNING(( "expected fd_sandbox_setup_mountfs to exit normally: " "exited with status, status=%d", WEXITSTATUS(wstatus) ));
+      return -1;
+    } else if ( FD_LIKELY ( WIFSIGNALED( wstatus ) ) ) {
+      FD_LOG_WARNING(( "expected fd_sandbox_setup_mountfs to exit normally: " "killed by signal %s", strsignal(WTERMSIG(wstatus)) ));
+      return 0;
+    } else if ( FD_UNLIKELY ( WIFSTOPPED( wstatus ) ) ) {
+      FD_LOG_WARNING(( "expected fd_sandbox_setup_mountfs to exit normally: " "stopped by signal %s", strsignal(WSTOPSIG(wstatus)) ));
+      return -1;
+    } else {
+      FD_LOG_WARNING(( "unexpected termination: wstatus(%d)", wstatus ));
+      return -2;
+    }
+    return -1;
+
+  } else { // child
+    fd_sandbox_setup_mountns(NULL);
+
+    struct dirent * entry;
+    DIR *dir = opendir("/");
+    if ( FD_UNLIKELY ( !dir ) ) {
+      FD_LOG_WARNING(( "open: expected to work on `/`: %s", strerror(errno) ));
+      return -1;
+    }
+
+    // Ensure that directory is empty
+    while ((entry = readdir(dir))) {
+      if ( FD_UNLIKELY ( strcmp( ".", entry->d_name ) && strcmp( "..", entry->d_name ) ) ) {
+        FD_LOG_WARNING(( "root directory is not empty" ));
+        exit(1);
+      }
+    }
+
+    // Ensure that `/` is owned by root
+    struct stat sb;
+    if ( FD_UNLIKELY ( stat( "/", &sb ) )) {
+      FD_LOG_WARNING(( "stat: expected to work on `/`: %s", strerror(errno) ));
+      exit(1);
+    }
+
+    if ( FD_UNLIKELY ( sb.st_uid ) ) {
+      FD_LOG_WARNING(( "root directory should be owned by user root but is owned by uid=%d", sb.st_uid ));
+      exit(1);
+    }
+
+    if ( FD_UNLIKELY ( sb.st_gid ) ) {
```

Comment:
If I understand this test correctly, it is just checking the truthiness of the `st_uid` and `st_gid` of the statbuf. Does it make sense to be explicit and check for `sb.st_uid == 0` and `sb.st_gid == 0`?

Reply:
https://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdbool.h.html

It's just checking if it's 0 as 0 is the only "false" value. IIRC, there are a lot of other places in FD where comparisons like this are made so I just consider it idiomatic. I'll defer to @ripatel-jump to know if the correction is needed.

Reply:
That works. I just like the python approach of "explicit is better than implicit", though I'll always defer to whatever is idiomatic for the language and project. Likely no correction at all.